### ISSUE TITLE
DSpellCheck on Windows 10 issue

### ISSUE THREAD

**Issue by donho** [notepad-plus-plus/notepad-plus-plus#427]
Here is what I got from MS for Notepad++ issue on Windows 10:

> Dear Windows Developer,
> 
> I’m a program manager in the Ecosystem Readiness Team at Microsoft. We work with Microsoft’s test organization to notify developers when compatibility issues have been identified in their applications during routine internal testing. We’re reaching out to notify you of a potential issue in one of your applications.
>  
> The issue details and suggested resolution are below. Our goal is to work with you to address this issue and to understand what your expected timeline to address this issue might be. If you have any questions about the details below or have already addressed this issue in a forthcoming update, please let me know.
                                                                                                                                                               

> Publisher
> Don Ho

> Product
> Notepad++

> Reference #
> 2836903

> Issue / Repro Steps
> APPLICATION_HANG_BusyHang_cfffffff_DSpellCheck.dll ! Unknown
 
> NOTE: This issue is observed in Windows builds of Windows 10
 
> Description:
> Approximately 612 users have experienced a crash on their Windows 10 machine inside of the application.
 
> Resolution:
> It is recommended that developers prepare for the release of Windows 10 by downloading the latest Insider Preview of the build and debugging their application against the new OS.
 
  
> Windows 10 & Visual Studio - https://dev.windows.com/en-US/downloads/windows-10-developer-tools?OCID=insider
> Forums - http://answers.microsoft.com/en-us/insider/forum?auth=1
 
 
> If you find that you need more information about the crash after local debugging, please reach out to us and we can help provide more information.
 
 
> Thank you for your continued support of Windows!

**Comment by Predelnik** [comment]
So I've checked on windows 10 build 10166 with Notepad++ 6.7.9.1 on VM it seems like it isn't obviously reproducible at least, I'll continue looking into it.

**Comment by ariccio** [comment]
It's really cool that *Microsoft notified you* about this problem.

**Comment by dartraiden** [comment]
I encountered this bug in Windows 8.1. N++ periodically (but not always) hangs when closing a window. 
This is not even required to edit the file, it was enough to open it and try to close. Removing DSpellCheck solve the problem.

**Comment by Rikk** [comment]
@donho What do you think about detecting Windows 10 in installer and don't install DSpellCheck only in this case? It is a great plugin, people surely is missing it in other Windows versions. Never had a crash due to it (W7-8.1); the only issue I recall is the "find in files", which is long fixed.

Btw, another one, [HexEditor, may be crashing in W10](https://notepad-plus-plus.org/community/topic/267/notepad-in-windows-10-wont-open-nfo-files) too.

**Comment by Predelnik** [comment]
@Rikk I'm now using Windows 10 on regular basis, haven't encountered any specific problems yet. Have no idea why it disappeared from installation.

@dartraiden I've put version 1.2.14.2 onto N++ plugins admin website for a while already, looks like plugin manager admins haven't updated it's database for a while.

**Comment by Rikk** [comment]
> Have no idea why it disappeared from installation.

Well, unfortunately, it is due to the first post on this issue, simplifying: Donho received a machine-written notification direct from Microsoft complaining of 612 crashes on a pre-release Win 10 version, related to the plugin. Error address is just this:
`APPLICATION_HANG_BusyHang_cfffffff_DSpellCheck.dll ! Unknown`

**Comment by Predelnik** [comment]
Well since a lot of W10 preview users downloaded it from the website probably, it's possible that it was 6.7.9 version which appeared to be broken but wasn't put on auto-update. I just have not a slightest idea what could be affected due to changes in Windows 10 and from my experience it seems like everything works the same as before.

**Comment by cpasmoi** [comment]
i use win10 and there is no problem, it was the old version of dspell which was hanging.

**Comment by mkitchin** [comment]
Hi there,

Just noticed this issue and that spellcheck had been disabled in a recent notepad++ release. FWIW, I have Win10 release (x64) and the described problem did occur once I re-installed dspellcheck from the plugin manager. Specifically, after closing tabs and then clicking the close box on notepad++, the notepad++ UI will freeze and require a kill. No dspellcheck plugin, no freeze.

I found the notepad++ spellcheck feature to be super-useful (and bizarrely so, since I use it as a source code editor most of the time), so it would be great if this functionality gets restored, soon.

Thanks again for your efforts.

**Comment by Predelnik** [comment]
@mkitchin if you downloaded it from plugin manager it's probably 1.2.14.0 version, the problem most likely would be resolved with installing 1.2.14.2 https://github.com/Predelnik/DSpellCheck/releases/tag/1.2.14.2

Maybe I should write to plugin manager admin to hurry up the update process, because even I myself run into this problem recently, once.

**Comment by dartraiden** [comment]
>> write to plugin manager admin
https://sourceforge.net/u/davegb3/profile/
I sent him a private message on sourceforge more than a week ago

**Comment by Predelnik** [comment]
@dartraiden Thank you for that. I've sent him e-mail now just in case. Hope it'll be resolved.

**Comment by mkitchin** [comment]
@Predelnik Thanks for the advice and the link. I obtained the latest release and the problem has gone away. It was pretty easy to repro and completely consistent, before, so I'd say this was a winner.

**Comment by JVimes** [comment]
Once the Plugin Manager is serving the fixed version, it would be good to make the installer warning more helpful, "... Please install the latest version of the Plugin from the Plugin Manager". 

I haven't hit the bug, myself, and was confused that installing the disabled Plugin worked fine for me.

**Comment by Kadah** [comment]
The version of DSpellCheck I had from the plugin manager was 1.2.14.1.

Would have nicer if the message from the installer about removing it stated that it was only a Win10 issue.

**Comment by dartraiden** [comment]
I have experienced this problem on windows 8.1. Microsoft discovered this problem due to the abundance of telemetry in the beta versions of Windows 10 (they reported the problem before release).

**Comment by taras-b** [comment]
As another N++ plugin developer I do know the pain of not being able to deliver the fix to your users for months. The fact that the plugin manager DB has been updated is irrelevant as it's only pushed to the users every few months or so. Thus it could take only 10 mins to make the fix but i's delivery will depend completely on the Plugin Manager team schedule.

I was able to solve the problem completely by providing an alternative channel for updated. The plugin does the checking for updates every day (configurable) and the about box allows user to check and install the updates at any time:


![image](https://cloud.githubusercontent.com/assets/3348045/10296818/1d5df240-6c18-11e5-8f99-2f1ba10d5a8b.png)


You can (if you wish) use the same solution. The plugin code is on [CodePLex](https://csscriptnpp.codeplex.com). Help yourself out.:) 

BTW, great job with your plugin!

**Comment by einst3** [comment]
i have found this issue on win7-win10 on the install version of npp. it will also corrupt some plugin settings for portable npp when it happens. i wish there were a good spell check for npp. if there is one please let me know.

**Comment by taras-b** [comment]
As it is stated in this thread the problem is solved in v1.2.14.2.
But you need to download and install the binaries manually from here: https://github.com/Predelnik/DSpellCheck/releases/tag/1.2.14.2

It's really the PluginManager who is 'preventing' the fix from hitting the user's desktops.

**Comment by TBugReporter** [comment]
It seems the code to disable DSpellCheck also does not discriminate regarding DSpellCheck versions, let alone discriminating Windows versions (for example, this machine I'm on now still runs W7, and therefore has no reason to have DSpellCheck disabled).  I installed 1.2.14.2 a while ago on all our machines; in the intervening time, I've also downloaded and installed several updated versions of Notepad++.  Each time, DSpellCheck was unceremoniously (and indiscriminatingly) disabled, and frankly, I'm getting sick of it.  Considering that this issue was first disclosed almost six months ago, I would think that the involved parties have had more than enough time to resolve this; what's still holding this up?

**Comment by dartraiden** [comment]
Seems like problem already non-actual. PluginManager offers DSpellCheck 1.2.14.2 which already do not have this error. So we can return plugin into package and we no longer need to turn it off when install a new version of the N ++.

**Comment by taras-b** [comment]
> Considering that this issue was first disclosed almost six months ago, I would think that the involved parties have had more than enough time to resolve this; what's still holding this up?

Being plugin developer by myself I can put my 5 cents here. 

It's very unlikely it's due to the DSpellCheck author lack of trying. I know that he fixed the problem (which wasn't even his) very quickly and for months and months PluginManager was still pushing the older incompatible binaries. And now it looks like DSpellCheck is smeared as "unstable" forever. Actually there is no mechanism for reviewing/reverting this verdict. It's completely up to PluginManager team to clear previously discriminated individual plugins.

**Comment by xylographe** [comment]
@TBugReporter 
Just rename `plugins\DSpellCheck.dll`, for example to `DSpellCheckOK.dll`, and it won't be disabled.

**Comment by Rikk** [comment]
>It's completely up to PluginManager team to clear previously discriminated individual plugins.

No, it's all up to @donho to include updated and fixed version in installer. Npp is his personal toy.

**Comment by donho** [comment]
@taras-b I think it'll be better that Plugin Manager's DB is on Github, but I have to contact Dave Brotherstone (Plugin Manager's author) for this issue. 
@Rikk  The fact I don't accept your buggy PR doesn't mean that Notepad++ is my personal toy.

**Comment by Rikk** [comment]
I'm not even looking at Npp anymore. But I saw you just accepted my fix to border=0. Thanks, and other users will also say thanks. And yes, you treat it as your toy. #fact. End of history. Goodbye.

**Comment by taras-b** [comment]
Just trying to defuse the conversation :)

>No, it's all up to @donho to include updated and fixed version in installer

But right now N++ is distributed with the FIXED version of the plugin. It's just marked in the PluginManager as unstable. It is really a single matter of establishing the process of Markin/Unmarking plugins as unstable in the Plugin manager. May be it should be done per plugin version.

**Comment by donho** [comment]
@Rikk If it makes you feel better that I "accepted you fix", then it is. :)

**Comment by donho** [comment]
@taras-b By "Plugin Manager's DB is on Github", I mean that there'll be more contributors to review the plugins, so @bruderstein won't be alone to process all the requests.

**Comment by taras-b** [comment]
If you are saying that in this case keeping a stable plugin marked as unstable is just as a single miss then I have no problems accepting that. After all I see no evidence that it happens all the time. Thus I see no issue here.  

Though I do find it inconvenient (at least) that delivering the fix could take up to 3-4 month after the fix is made public. And the plugin developer has no control over it. 

Don't get me wrong I perfectly understand  why the plugin manager config file is not updated more frequently. I am just stating that the current model pops some questions about the developer experiences. 

In my case I recognized this problem earlier and implemented plugin's own internal update channel. I just had no other choice as my plugin could not afford "slow releases" model. But I don't really see any good GENERIC solution (or work around) that the all pluging developers would benefit from.

Have a nice day.

**Comment by TBugReporter** [comment]
@xylographe:
> Just rename `plugins\DSpellCheck.dll`, for example to `DSpellCheckOK.dll`, and it won't be disabled.

Yes, this is a simple workaround, but one of my points was that after six months, it seems ridiculous that users still need to work around an issue that's no longer an issue - except for cleaning up after the previous half-baked workarounds.

**Comment by bruderstein** [comment]
Hi, plugin manager author here.

TL;DR:  I'm sorry. If you want to help you can. It's going to get better by or in Janauary.

Firstly, I apologise for the frustration I've caused. The releases are still a fair bit of work, as each update needs to be checked, and around 40% generally need correcting.  I appreciate that's probably due to it not being super-easy to test the changes. It's actually not too bad, just download the testing version and add two lines in your PluginManager.ini, but a lot of plugin devs don't seem to do it. Probably because they know it won't be released quickly.

If anyone from the community wants to be added to the people that can do the releases, I'm more than happy to walk someone through the process, and provide my scripts etc that I use. Please reply to this issue you want to take this on.

To address the point of a github hosted list - we did that originally, and whilst I did receive a few pull requests, it was significantly more work, and I think there were only about 6 people that contributed - that from a base of over a hundred plugin developers. Today, pretty much all changes are made by the plugin developers themselves.  

My primary concern currently is getting off sourceforge. I've written a new "plugins" server, and lined up a sponsorship deal for the hosting (I've talked before about this, but in a month with a release, the tiny plugins zip creates about 800GB in traffic).  I've spoken to Github, and they wanted around $80 per month, which quite frankly, I'm not prepared to pay.

The new server should be live by/in January. This is specifically designed to ease the transition to a model where changes are automatically pushed to live, once they've been checked.  

Once that's live, I'll add a "push live" button to the admin system for admins, so that people can easily push all changes live. Unfortunately, with the current setup, this cannot be available for everybody, as all plugins must be published at the same time. But, with 2 or 3 "admins", this should be a much quicker process.  Until that goes live, as I said earlier, anyone who wants to help with the current setup is more than welcome to.

The slightly longer term plan, which I've also talked about before, is to have an two-stage admin system, that individual plugin changes can be "approved", and then they are automatically pushed live at regular intervals.  That project is going quite well, but I admit it's fairly low priority at the moment.

If you've got  any comments on this, or wish to become an admin, please add a comment with your plugin admin system username, and we'll sort out a time to walk through the process.

**Comment by bruderstein** [comment]
One point I forgot - the stability ranking in plugin manager can be changed by anyone. I have on very rare occasions set it due to multiple reports of crashes. The author is free to set it back after fixing the issue(s).

**Comment by taras-b** [comment]
Thank you Dave for such a detailed explanation. 
>The author is free to set it back after fixing the issue(s).

This actually completely changes everything. I missed this one. My bad. It means that in fact there is a simple mechanism for clearing the plugins that were previously marked as unstable.

With the new system in place will there be any changes for the submitting the updates. Currently I am using http://www.brotherstone.co.uk/npp/pm/admin web interface for this.Is it going to stay the same?

**Comment by Predelnik** [comment]
Thanks to Dave for finally updating database and explanation, maybe new system will make things better.

Personally - while removal from installation discouraged me to make any big changes in plugin, I continue to use it as a user (half of the time on windows 10) and also get annoyed each time I have to install N++ update.

**Comment by bruderstein** [comment]
@taras-b - The URL will stay the same, at least initially. 

Stage 1 will just be an extra option for anyone that wants to help with a "publish" option, which will publish the entire plugins.xml as live.  Obviously this can't be for everyone, as whoever pushes that button needs to have tested all the updates for all the plugins since the last publish.

Stage 2 will be a new interface, hopefully with the couple of bits that are often missed/forgotten in the current interface made more obvious.  The idea will remain the same, but the interface might be a bit different. It's possible the URL will change, but I'll publicise that at the time, and create an automatic redirect from the old address.

**Comment by TBugReporter** [comment]
I just upgraded to Notepad++ 6.8.8 yesterday, and for the first time in several months, I *haven't* had to deal with the installer disabling DSpellCheck.  I don't know who's responsible for this change, but to whoever it is:  Thank you!

**Comment by vgturtle127** [comment]
@donho This is still disabled in the latest 6.9 release, I thought you disabled the removal of DSpellcheck?

@bruderstein I will help contribute in any way I can if you will just tell me what to do and link to your repository and stuff. :smile:

**Comment by johnrb68** [comment]
So many many months later, and this STILL is not fixed????

When is this going to work properly again???

**Comment by alexjbs** [comment]
You have the option to whether disable it or not.
Even if you disable it, later you can open the Plugin Manager, select DSpellCheck in the Available tab and click Install.

**Comment by un1versal** [comment]
You have the option to whether disable it or not.
Even if you disable it, later you can open the Plugin Manager, select DSpellCheck in the Available tab and click Install.

If only it were that simple, It wont install, it fails. just copying it back from disabled or manually downloading it will work.

**Comment by codewizard13** [comment]
I am a developer in the USA. It is too bad this plugin is not working. There should be a notation on this plugin in the Plugin Manager indicating that it doesn't work yet and "use at your own risk". And I'm talking about something obvious like the name "DSpellCheck" grayed out and the warning I mentioned in the same row or column in red text.

**Comment by thewriteway** [comment]
is this going to be fixed ever?

**Comment by JVimes** [comment]
It's working on my machines. I'm curious what's happening for others. I installed Notepad++ (via Chocolatey), then installed DSpellCheck via plugin manager. I had to go into the plugin settings and "download dictionaries" or no spell checking. No errors/warnings and the plugin works fine. Would be nice if it came working with Notepad++.

**Comment by oleg-shilo** [comment]
I second it. No problems whatsoever. 

Since v1.2.14.2 release (more than a year ago) the plugin address the issue completely (at least for me). And now PluginManager also deploys the very same release so you don't have to pull it from GitHub. 

May be the lack of pre-deployed dictionaries on fresh install is putting people off...

**Comment by thewriteway** [comment]
Should have tried upgrading first as I have now just installed Notepad++ V7, installed DSpellCheck and its working perfectly. :) Thanks for the follow up JVimes & oleg-shilo.

**Comment by JVimes** [comment]
Glad to help :) Ya, the plugin _really_ needs to come with some dictionaries or it seems broken. I wasn't sure if that was happening for everyone, or I was doing something weird.
